pkg/dataobj/logs_reader.go

```diff
@@ -109,33 +113,77 @@ func (r *LogsReader) Read(ctx context.Context, s []Record) (int, error) {
 		}
 	}
 
-	r.buf = slices.Grow(r.buf, len(s))
+	r.buf = slicegrow.GrowToCap(r.buf, len(s))
 	r.buf = r.buf[:len(s)]
 
+	// Fill the row buffer with empty values so they can re-use the memory we pass in.
+	for i := range r.buf {
+		r.buf[i].Values = slicegrow.GrowToCap(r.buf[i].Values, len(r.columns))
+		r.buf[i].Values = r.buf[i].Values[:len(r.columns)]
+		for j := range r.buf[i].Values {
+			if r.buf[i].Values[j].IsNil() {
+				r.buf[i].Values[j] = dataset.ByteArrayValue(make([]byte, 0, 64))
+			}
+		}
+	}
+
 	n, err := r.reader.Read(ctx, r.buf)
 	if err != nil && !errors.Is(err, io.EOF) {
 		return 0, fmt.Errorf("reading rows: %w", err)
 	} else if n == 0 && errors.Is(err, io.EOF) {
 		return 0, io.EOF
 	}
 
+	metadataColumns := 0
+	for _, column := range r.columnDesc {
+		if column.Type == logsmd.COLUMN_TYPE_METADATA {
+			metadataColumns++
+		}
+	}
+
+	// Pre-allocate memory for metadata
+	for i := range s {
```

Comment:
nit(unrelated, future): It looks like we're initializing metadata columns even if they may be unused. This can (a) likely be minimized so we only pay initialization costs when necessary, this time on the `s []Record` argument instead of the internal `buf`. This is a bit more complicated because caching that would likely require a type wrapper over `[]Record`. And (b) it looks like we're initializing capacity based on the total metadata columns (which can be very large), rather than those requested by the query.

Neither optimization needs to be applied today, but it looks like there's some significant opportunity here.